client-apis/microprofile-restclient/src/main/java/org/wildfly/swarm/microprofile/restclient/InvocationContextImpl.java

```diff
@@ -0,0 +1,168 @@
+
+    private final List<InterceptorInvocation> chain;
+
+    /**
+     * @param target
+     * @param method
+     * @param args
+     * @param chain
+     */
+    InvocationContextImpl(Object target, Method method, Object[] args, List<InterceptorInvocation> chain) {
+        this.target = target;
+        this.method = method;
+        this.args = args;
+        this.contextData = new HashMap<>();
+        this.position = 0;
+        this.chain = chain;
+    }
+
+    boolean hasNextInterceptor() {
+        return position < chain.size();
+    }
+
+    protected Object invokeNext() throws Exception {
+        int oldPosition = position;
+        try {
+            return chain.get(position++).invoke(this);
+        } finally {
+            position = oldPosition;
+        }
+    }
+
+    protected Object interceptorChainCompleted() throws Exception {
+        try {
+            return method.invoke(target, args);
+        } catch (InvocationTargetException e) {
+            if (e.getCause() instanceof ResponseProcessingException) {
+                ResponseProcessingException rpe = (ResponseProcessingException) e.getCause();
+                // Note that the default client engine leverages a single connection
+                // MP FT: we need to close the response otherwise we would not be able to retry if the method returns javax.ws.rs.core.Response
+                rpe.getResponse().close();
+                Throwable cause = rpe.getCause();
+                if (cause instanceof RuntimeException) {
+                    throw (RuntimeException) cause;
+                }
+            }
+            throw e;
+        }
+    }
+
+    @Override
+    public Object proceed() throws Exception {
+        try {
+            if (hasNextInterceptor()) {
+                return invokeNext();
+            } else {
+                return interceptorChainCompleted();
+            }
+        } catch (InvocationTargetException e) {
+            Throwable cause = e.getCause();
+            if (cause instanceof Error) {
+                throw (Error) cause;
+            }
+            if (cause instanceof Exception) {
+                throw (Exception) cause;
+            }
+            throw new RuntimeException(cause);
+        }
+    }
+
+    @Override
+    public Object getTarget() {
+        return target;
+    }
+
+    @Override
+    public Method getMethod() {
+        return method;
+    }
+
+    @Override
+    public Constructor<?> getConstructor() {
+        return null;
+    }
+
+    @Override
+    public Object[] getParameters() throws IllegalStateException {
+        return args;
+    }
+
+    @Override
+    public void setParameters(Object[] params) throws IllegalStateException, IllegalArgumentException {
+        this.args = params;
+    }
+
+    @Override
+    public Map<String, Object> getContextData() {
+        return contextData;
+    }
+
+    @Override
+    public Object getTimer() {
+        return null;
```

Comment:
Maybe throw `IllegalStateException`?

Reply:
I don't think so, `getTimer()` method's javadoc: _"returns the timer object, or null if the invocation did not apply to a timeout method"_.

Reply:
Ah OK, sorry -- as I said, I don't really know anything about interceptors :-)